File: __tests__/lib/progression-advice.test.ts
// Tests du conseil de progression muscu (lib/progression-advice.ts) : mapping du
// ressenti vers un score, heuristique augmente/maintiens/réduis, libellés. Pur.
import {
  adviceLabel,
  difficultyLabel,
  difficultyScore,
  suggestProgression,
} from '@/lib/progression-advice';

describe('difficultyScore', () => {
  it('mappe chaque ressenti sur un score ordonnable, null sinon', () => {
    expect(difficultyScore('facile')).toBe(1);
    expect(difficultyScore('moyen')).toBe(2);
    expect(difficultyScore('dur')).toBe(3);
    expect(difficultyScore(null)).toBeNull();
    expect(difficultyScore(undefined)).toBeNull();
  });
});

describe('difficultyLabel', () => {
  it('libelle chaque ressenti en français', () => {
    expect(difficultyLabel('facile')).toBe('Facile');
    expect(difficultyLabel('moyen')).toBe('Moyen');
    expect(difficultyLabel('dur')).toBe('Dur');
  });
});

describe('suggestProgression', () => {
  it('aucune note exploitable : maintiens', () => {
    expect(suggestProgression([])).toBe('maintiens');
    expect(suggestProgression([null, undefined])).toBe('maintiens');
  });

  it('deux séances faciles : augmente', () => {
    expect(suggestProgression(['facile', 'facile'])).toBe('augmente');
  });

  it('dernière facile sans dur récent : augmente', () => {
    expect(suggestProgression(['moyen', 'facile'])).toBe('augmente');
  });

  it('dernière dure : réduis (priorité sur la moyenne)', () => {
    expect(suggestProgression(['facile', 'dur'])).toBe('reduis');
  });

  it('moyen dominant : maintiens', () => {
    expect(suggestProgression(['moyen', 'moyen', 'moyen'])).toBe('maintiens');
  });

  it('deux séances dures : réduis', () => {
    expect(suggestProgression(['dur', 'dur'])).toBe('reduis');
  });

  it('ignore les séances non notées', () => {
    expect(suggestProgression([null, 'facile', null, 'facile'])).toBe('augmente');
  });

  it('respecte la fenêtre : un « dur » ancien sort du calcul', () => {
    // window=2 : seules ['facile','facile'] comptent -> augmente.
    expect(suggestProgression(['dur', 'facile', 'facile'], 2)).toBe('augmente');
    // window=3 : le « dur » revient dans la fenêtre, dernière reste facile mais
    // un « dur » est présent -> on retombe sur la moyenne (1+1+3)/3 ≈ 1.67 -> maintiens.
    expect(suggestProgression(['dur', 'facile', 'facile'], 3)).toBe('maintiens');
  });
});

describe('adviceLabel', () => {
  it('formule chaque conseil au tutoiement', () => {
    expect(adviceLabel('augmente')).toBe('Augmente les reps ou la charge');
    expect(adviceLabel('reduis')).toBe('Réduis la charge, soigne la technique');
    expect(adviceLabel('maintiens')).toBe('Maintiens, charge bien calibrée');
  });
});

File: app/src/test/java/ovh/battistella/elan/domain/ProgressionAdviceTest.kt
// Tests du conseil de progression muscu (ProgressionAdvice.kt) : mapping du
// ressenti vers un score, heuristique augmente/maintiens/réduis, libellés. Pur.
package ovh.battistella.elan.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class ProgressionAdviceTest {
    @Test
    fun `difficultyScore mappe chaque ressenti sur un score ordonnable, null sinon`() {
        assertEquals(1, difficultyScore(Difficulty.FACILE))
        assertEquals(2, difficultyScore(Difficulty.MOYEN))
        assertEquals(3, difficultyScore(Difficulty.DUR))
        assertNull(difficultyScore(null))
    }

    @Test
    fun `difficultyLabel libelle chaque ressenti en français`() {
        assertEquals("Facile", difficultyLabel(Difficulty.FACILE))
        assertEquals("Moyen", difficultyLabel(Difficulty.MOYEN))
        assertEquals("Dur", difficultyLabel(Difficulty.DUR))
    }

    @Test
    fun `aucune note exploitable - maintiens`() {
        assertEquals(ProgressionAdvice.MAINTIENS, suggestProgression(emptyList()))
        assertEquals(ProgressionAdvice.MAINTIENS, suggestProgression(listOf(null, null)))
    }

    @Test
    fun `deux séances faciles - augmente`() {
        assertEquals(ProgressionAdvice.AUGMENTE, suggestProgression(listOf(Difficulty.FACILE, Difficulty.FACILE)))
    }

    @Test
    fun `dernière facile sans dur récent - augmente`() {
        assertEquals(ProgressionAdvice.AUGMENTE, suggestProgression(listOf(Difficulty.MOYEN, Difficulty.FACILE)))
    }

    @Test
    fun `dernière dure - réduis (priorité sur la moyenne)`() {
        assertEquals(ProgressionAdvice.REDUIS, suggestProgression(listOf(Difficulty.FACILE, Difficulty.DUR)))
    }

    @Test
    fun `moyen dominant - maintiens`() {
        assertEquals(
            ProgressionAdvice.MAINTIENS,
            suggestProgression(listOf(Difficulty.MOYEN, Difficulty.MOYEN, Difficulty.MOYEN)),
        )
    }

    @Test
    fun `deux séances dures - réduis`() {
        assertEquals(ProgressionAdvice.REDUIS, suggestProgression(listOf(Difficulty.DUR, Difficulty.DUR)))
    }

    @Test
    fun `ignore les séances non notées`() {
        assertEquals(
            ProgressionAdvice.AUGMENTE,
            suggestProgression(listOf(null, Difficulty.FACILE, null, Difficulty.FACILE)),
        )
    }

    @Test
    fun `respecte la fenêtre - un dur ancien sort du calcul`() {
        val recent = listOf(Difficulty.DUR, Difficulty.FACILE, Difficulty.FACILE)
        // window=2 : seules [facile, facile] comptent -> augmente.
        assertEquals(ProgressionAdvice.AUGMENTE, suggestProgression(recent, 2))
        // window=3 : le « dur » revient dans la fenêtre, dernière reste facile mais
        // un « dur » est présent -> on retombe sur la moyenne (1+1+3)/3 ≈ 1.67 -> maintiens.
        assertEquals(ProgressionAdvice.MAINTIENS, suggestProgression(recent, 3))
    }

    @Test
    fun `fenêtre nulle - toutes les séances notées comptent (slice(-0) JS)`() {
        // [dur, facile, facile] entier : dernière facile mais « dur » présent -> moyenne 1,67 -> maintiens.
        assertEquals(
            ProgressionAdvice.MAINTIENS,
            suggestProgression(listOf(Difficulty.DUR, Difficulty.FACILE, Difficulty.FACILE), 0),
        )
    }

    @Test
    fun `adviceLabel formule chaque conseil au tutoiement`() {
        assertEquals("Augmente les reps ou la charge", adviceLabel(ProgressionAdvice.AUGMENTE))
        assertEquals("Réduis la charge, soigne la technique", adviceLabel(ProgressionAdvice.REDUIS))
        assertEquals("Maintiens, charge bien calibrée", adviceLabel(ProgressionAdvice.MAINTIENS))
    }
}
